libs/runtime/ebpf_epoch.c

```diff
@@ -324,95 +321,73 @@ ebpf_epoch_terminate()
-        // After the semaphore is acquired, then there is at least 1 available epoch entry.
-    }
-
-    // If the current thread is not preemptible, then there is at least 1 available epoch entry.
-
-    // Grab the CPU lock.
-    ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_epoch_cpu_table[current_cpu].lock);
+    epoch_state->irql_at_enter = _ebpf_epoch_raise_to_dispatch_if_needed();
+    epoch_state->cpu_id = ebpf_get_current_cpu();
 
-    // Get the next available epoch entry.
-    epoch_entry = _ebpf_epoch_next_available_epoch_entry(&_ebpf_epoch_cpu_table[current_cpu]);
-    ebpf_assert(epoch_entry != NULL);
-    _Analysis_assume_(epoch_entry != NULL);
-    ebpf_assert(epoch_entry->epoch == 0);
+    ebpf_epoch_cpu_entry_t* cpu_entry = &_ebpf_epoch_cpu_table[epoch_state->cpu_id];
+    epoch_state->epoch = cpu_entry->current_epoch;
+    ebpf_list_insert_tail(&cpu_entry->epoch_state_list, &epoch_state->epoch_list_entry);
 
-    // Capture the current epoch.
-    epoch_entry->epoch = _ebpf_current_epoch;
-    epoch_entry->is_preemptible = is_preemptible;
-
-    // Release the CPU lock.
-    ebpf_lock_unlock(&_ebpf_epoch_cpu_table[current_cpu].lock, state);
-    return epoch_entry;
+    _ebpf_epoch_lower_to_previous_irql(epoch_state->irql_at_enter);
 }
-
-void
+#pragma warning(pop)
+
+#pragma warning(push)
+#pragma warning( \
+    disable : 28166) // warning C28166: The function 'ebpf_epoch_exit' does not restore the IRQL to the value that was
+                     // current at function entry and is required to do so. IRQL was last set to 2 at line 353.
+// the IRQL.
+_IRQL_requires_same_ void
 ebpf_epoch_exit(_In_ ebpf_epoch_state_t* epoch_state)
 {
-    uint32_t current_cpu = _ebpf_epoch_get_cpu_id_from_state(epoch_state);
-    // Capture preemptible state outside lock
-    bool is_preemptible = epoch_state->is_preemptible ? true : false;
-    bool release_free_list = false;
+    KIRQL old_irql = _ebpf_epoch_raise_to_dispatch_if_needed();
 
-    ebpf_assert(is_preemptible == ebpf_is_preemptible());
-    ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_epoch_cpu_table[current_cpu].lock);
+    // Assert the IRQL is the same as when ebpf_epoch_enter() was called.
+    ebpf_assert(old_irql == epoch_state->irql_at_enter);
 
-    // Mark the epoch state as available.
-    ebpf_assert(epoch_state->epoch != 0);
-    epoch_state->epoch = 0;
-    epoch_state->is_preemptible = false;
+    uint32_t cpu_id = ebpf_get_current_cpu();
 
-    // Mark the epoch state as not stale.
-    _ebpf_epoch_cpu_table[current_cpu].stale = false;
+    // Special case: Thread has moved to a different CPU since entering the epoch.
+    if (cpu_id != epoch_state->cpu_id) {
+        // Assert that the IRQL is < DISPATCH_LEVEL. If it is DISPATCH_LEVEL, then the thread moved to a different CPU
+        // (by dropping below DISPATCH_LEVEL) and calling ebpf_epoch_exit(). This is not allowed.
+        ebpf_assert(epoch_state->irql_at_enter < DISPATCH_LEVEL);
```

Comment:
Would it be safe to continue if this case happens in retail?

Reply:
No, we should probably bugcheck if this happens.

Reply:
https://github.com/microsoft/ebpf-for-windows/issues/2827